res/res_pjsip_refer.c

```diff
@@ -786,6 +791,499 @@ static void refer_blind_callback(struct ast_channel *chan, struct transfer_chann
 		ast_channel_unlock((session)->channel);											\
 	} while (0)																			\
 
+struct refer_data {
+	struct ast_refer *refer;
+	char *destination;
+	char *from;
+	char *refer_to;
+	int to_self;
+};
+
+static void refer_data_destroy(void *obj)
+{
+	struct refer_data *rdata = obj;
+
+	ast_free(rdata->destination);
+	ast_free(rdata->from);
+	ast_free(rdata->refer_to);
+
+	ast_refer_destroy(rdata->refer);
+}
+
+static struct refer_data *refer_data_create(const struct ast_refer *refer)
+{
+	char *uri_params;
+	const char *destination;
+	struct refer_data *rdata = ao2_alloc(sizeof(*rdata), refer_data_destroy);
```

Comment:
Does this need to be allocated with a lock?

Reply:
Actually, I don't think so.

- The `refer_data` is only used in `sip_refer_send()`, which pushes `refer_send()` to the task serializer and passes the `refer_data` to it.
- In `refer_send()` this `refer_data` is attached to the `pjsip_dialog` with `pjsip_dlg_add_usage`.
- The `refer_out_of_dialog_module` accesses this only `on_tx_request` in `refer_on_tx_request()` with `pjsip_dlg_get_mod_data`, which is in a PJSIP thread, but I don't think there should be a race condition here.

So I think we can turn this into `ao2_alloc_options(sizeof(*rdata), refer_data_destroy, AO2_ALLOC_OPT_LOCK_NOLOCK)`.

Reply:
Switched to NOLOCK.